src/main/java/jenkins/plugins/openstack/compute/TemporaryServer.java

```diff
@@ -0,0 +1,243 @@
+
+    private static final AtomicInteger nodeCounter = new AtomicInteger();
+
+    private SlaveOptions slaveOptions;
+    private transient JCloudsCloud cloud;
+
+    @DataBoundConstructor
+    public TemporaryServer(final SlaveOptions slaveOptions) {
+        this.slaveOptions = slaveOptions;
+    }
+
+    public @Nonnull SlaveOptions getRawSlaveOptions() {
+        if (cloud == null) throw new IllegalStateException("Owner not set properly");
+        return slaveOptions;
+    }
+
+    /**
+     * Provision and connect as a slave.
+     *
+     * The node is to be added to Jenkins by the caller. At the time the method completes, the node is ready to be launched.
+     *
+     * @throws Openstack.ActionFailed Provisioning failed.
+     * @throws JCloudsCloud.ProvisioningFailedException Provisioning failed.
+     */
+    public @Nonnull JCloudsSlave provisionSlave(
+            @Nonnull JCloudsCloud cloud, @Nonnull ProvisioningActivity.Id id, TaskListener listener
+    ) throws JCloudsCloud.ProvisioningFailedException, InterruptedException {
+        int timeout = slaveOptions.getStartTimeout();
+        this.cloud = cloud;
+        Server nodeMetadata = provision(cloud);
+
+        JCloudsSlave node = null;
+        // Terminate node unless provisioned successfully
+        try {
+            node = new JCloudsSlave(id, nodeMetadata, null, slaveOptions);
+
+            String cause;
+            while ((cause = cloud.slaveIsWaitingFor(node)) != null) {
+                if ((System.currentTimeMillis() - node.getCreatedTime()) > timeout) {
+                    String timeoutMessage = String.format("Failed to connect agent %s within timeout (%d ms): %s", node.getNodeName(), timeout, cause);
+                    LOGGER.warning(timeoutMessage);
+                    throw new JCloudsCloud.ProvisioningFailedException(timeoutMessage);
+                }
+
+                Thread.sleep(2000);
+            }
+
+            return node;
+        } catch (Throwable ex) {
+            JCloudsCloud.ProvisioningFailedException cause = ex instanceof JCloudsCloud.ProvisioningFailedException
+                    ? (JCloudsCloud.ProvisioningFailedException) ex
+                    : new JCloudsCloud.ProvisioningFailedException("Unable to provision node: " + ex.getMessage(), ex)
+                    ;
+
+            if (node != null) {
+                // No need to call AbstractCloudSlave#terminate() as this was never added to Jenkins
+                node._terminate(TaskListener.NULL);
+            }
+            throw cause;
+        }
+    }
+
+    /**
+     * Provision OpenStack machine.
+     *
+     * @throws Openstack.ActionFailed In case the provisioning failed.
+     * @see #provisionSlave(JCloudsCloud, ProvisioningActivity.Id, TaskListener)
+     */
+    /*package*/ @Nonnull Server provision(@Nonnull JCloudsCloud cloud) throws Openstack.ActionFailed {
+        return provision(cloud, null);
+    }
+
+    @Restricted(NoExternalUse.class)
```

Comment:
This can be moved to a class level so we have the liberty to adjust the API later.